graphql_compiler/cost_estimation/interval.py

```diff
@@ -0,0 +1,91 @@
+import datetime
+from typing import Generic, Optional, TypeVar
+
+
+IntervalDomain = TypeVar("IntervalDomain", int, str, datetime.date, datetime.datetime)
+
+
+class Interval(Generic[IntervalDomain]):
+    def __init__(
+        self, lower_bound: Optional[IntervalDomain], upper_bound: Optional[IntervalDomain]
+    ) -> None:
+        """Initialize an interval. A value of None means unbounded."""
+        self.lower_bound: Optional[IntervalDomain] = lower_bound
+        self.upper_bound: Optional[IntervalDomain] = upper_bound
+
+    def is_empty(self) -> bool:
+        """Return whether the interval is empty"""
+        if self.lower_bound is None or self.upper_bound is None:
+            return False
+        return self.lower_bound > self.upper_bound
+
+    def __eq__(self, other) -> bool:
+        """Compare two intervals. Empty intervals are considered equal to each other."""
+        if self.is_empty() and other.is_empty():
+            return True
+        return self.lower_bound == other.lower_bound and self.upper_bound == other.upper_bound
+
+
+def measure_interval(interval: Interval[int]) -> Optional[int]:
+    if interval.lower_bound is None or interval.upper_bound is None:
+        return None
+    if interval.is_empty():
+        return 0
+    return interval.upper_bound - interval.lower_bound + 1
+
+
+def _get_stronger_lower_bound(lower_bound_a, lower_bound_b):
```

Comment:
Prefer type hints rather than `int or None` docstrings. I'm assuming this is copy-pasted from the other file.